sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala

```diff
@@ -328,3 +281,222 @@ object FrameBoundaryExtractor {
     case _ => None
   }
 }
+
+/**
+ * A window function is a function that can only be evaluated in the context of a window operator.
+ */
+trait WindowFunction extends Expression {
+  /** Frame in which the window operator must be executed. */
+  def frame: WindowFrame = UnspecifiedFrame
+}
+
+abstract class OffsetWindowFunction
+  extends Expression with WindowFunction with Unevaluable with ImplicitCastInputTypes {
+  val input: Expression
+  val default: Expression
+  val offset: Expression
+  val offsetSign: Int
+
+  override def children: Seq[Expression] = Seq(input, offset, default)
+
+  override def foldable: Boolean = input.foldable && (default == null || default.foldable)
+
+  override def nullable: Boolean = input.nullable && (default == null || default.nullable)
+
+  override lazy val frame = {
+    // This will be triggered by the Analyzer.
+    val offsetValue = offset.eval() match {
+      case o: Int => o
+      case x => throw new AnalysisException(
+        s"Offset expression must be a foldable integer expression: $x")
+    }
+    val boundary = ValueFollowing(offsetSign * offsetValue)
+    SpecifiedWindowFrame(RowFrame, boundary, boundary)
+  }
+
+  override def dataType: DataType = input.dataType
+
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(AnyDataType, IntegerType, TypeCollection(input.dataType, NullType))
```

Comment:
I think it does, returning any other type would violate the principle of least surprise. It is basically a coalesce of a value ```x``` in the leading/lagging row, and a ```default``` value in the current row.